### ISSUE TITLE
Add support to pull OCI kustomize manifests

### ISSUE THREAD

**Issue by NissesSenap** [kubernetes-sigs/kustomize#5134]
### Eschewed features

- [X] This issue is not requesting templating, unstuctured edits, build-time side-effects from args or env vars, or any other eschewed feature.

### What would you like to have added?

I would like to make it possible to build and pull OCI artifacts containing Kustomize manifests.

This is a feature that the flux community have implemented https://fluxcd.io/flux/installation, https://fluxcd.io/flux/cmd/flux_pull_artifact/

So a simple example, you can write the following and it will download the OCI artifact and unpack it.

```
flux pull artifact oci://ghcr.io/grafana-operator/kustomize/grafana-operator:v5.0.0-rc1 --output ./grafana-operator/
```

For those of us that uses flux it's nothing strange, but for none flux users it would be nice to be able to do the same using Kustomize.

So adding oci support to localize and build to be able to do something like this would be nice:

```
kustomize localize oci://ghcr.io/grafana-operator/kustomize/grafana-operator:v5.0.0-rc1
```

### Why is this needed?

OCI is getting a big foothold when it comes to artifact management in the CNCF landscape, and it will be used more and more. The possibility of signing and easily versioning manifests is something that should be possible through kustomize

### Can you accomplish the motivating task without this feature, and if so, how?

It's possible to solve this by using flux, but since it's a third party tool, I don't think the majority of kustomize users will want to do it.

### What other solutions have you considered?

N/A

### Anything else we should know?

It would also be nice to add support to create OCI artifact from manifests, I haven't thought of a potential good syntax.
As I see it, it could be done in two seperate features.
I would start with pulling, since OCI artifacts containing kustomize files already are available today.

I started this conversation in https://github.com/kubernetes-sigs/kustomize/issues/4996#issuecomment-1495925893 but I wanted an issue that I could reference. 

### Feature ownership

- [ ] I am interested in contributing this feature myself! 🎉

**Comment by NissesSenap** [comment]
Ping @natasha41575 I created a separate issue for this to make it a bit more visible.
I hope that was okay.

**Comment by natasha41575** [comment]
/triage accepted

I think we should support OCI for both kustomize build and kustomize localize.

Thanks for filing the issue!

**Comment by natasha41575** [comment]
Providing some more details on the requirements here:

`kustomize localize` and `kustomize build` should support the exact same feature set IMO. If we add oci support to `localize`, we also need to add it to `build`. 

That means that if we support localizing an oci manifest e.g. `kustomize localize oci://ghcr.io/grafana-operator/kustomize/grafana-operator:v5.0.0-rc1`, we _also) need to support building an oci manifest like `kustomize localize oci://ghcr.io/grafana-operator/kustomize/grafana-operator:v5.0.0-rc1`. 

Oci support also means that we support oci resources in the kustomization file. For example, we currently support:

```
# kustomization.yaml
resources:
- https://github.com/kubernetes-sigs/kustomize//examples/multibases?ref=v1.0.6
```

If we say we support oci in addition to git, we need to support:
```
resources:
- oci://ghcr.io/grafana-operator/kustomize/grafana-operator:v5.0.0-rc1
```

This case needs to be handled by both `kustomize build` and `kustomize localize`. 

Implementation-wise, both `localize` and `build` can use the same underlying code as it currently does for pulling remote git urls.

**Comment by FrenchBen** [comment]
@natasha41575 I went ahead and implemented the changes requested, however I couldn't find details around the `origins.originAnnotations` that you've mentioned
>The implementation for build also needs to ensure that origin annotations get added correctly when options.originAnnotations is set in the kustomization file, with corresponding tests added to api/krusty/originannotation_test.go.

Where can I see docs on the above?

**Comment by erikgb** [comment]
@FrenchBen Did you ever open a PR with your implementation? I am also very interested in this feature!

> Where can I see docs on the above?

I don't think this feature was ever documented. It seems to have been added by @natasha41575 in https://github.com/kubernetes-sigs/kustomize/pull/4065. From a user perspective, I expect this new feature to just set the `config.kubernetes.io/origin`  annotation in resources emitted from the OCI repository.

````
# kustomization.yaml
resources:
- oci://ghcr.io/grafana-operator/kustomize/grafana-operator:v5.0.0-rc1
buildMetadata: [originAnnotations]
````

Should probably add something like

````
metadata:
  annotations:
    config.kubernetes.io/origin: |
      path:  oci://ghcr.io/grafana-operator/kustomize/grafana-operator:v5.0.0-rc1/default/kube-prometheus-stack.yaml
````

to all resources emitted? See https://github.com/kubernetes-sigs/kustomize/blob/336bc14e1a1c5733564ffb6c9d33f0be04671ee1/api/krusty/originannotation_test.go

**Comment by NissesSenap** [comment]
Fancy seeing you here @erikgb :)
@FrenchBen have been working on the implementation, you can find the pr here https://github.com/kubernetes-sigs/kustomize/pull/5147

**Comment by petermicuch** [comment]
@natasha41575 - has this been put on hold? I've seen last relevant activity in the referenced pr [#5147](https://github.com/kubernetes-sigs/kustomize/pull/5147) back in august. It would be great to have this. I am handling dockers and helms as OCIs and am missing this for kustomize as well. This would allow me to handover complete app deployment composed of docker, helm and kustomize as a single reference to OCI registry which would be really cool.

**Comment by siegenthalerroger** [comment]
I'm not sure what the situation surrounding bounties/donations for specific features is here, but might it be possible in this case? I can see a lot of interest in this feature but a lack of progress with the open PR and would be willing to chip in personally and try and get my employer to do the same here.

**Comment by Clockwork-Muse** [comment]
I'm working on this (slowly) since I have reason to want this for myself.

**Comment by FrenchBen** [comment]
I hadn't seen much movement on the PR I had opened. 
Happy to revisit with @Clockwork-Muse

**Comment by Clockwork-Muse** [comment]
Making some progress.  
I've opened a draft PR if people want to look at it.  I currently have it working with OCI-layout resources (at least directories).
From the quick look over the tests, I'm not seeing anything that actually tests the remote git capabilities (probably mostly because it calls git itself).... but this code may need some remote testing capability...

**Comment by Clockwork-Muse** [comment]
.... The irony here is that I'm not using kustomize anymore - I need purge support for SSA-d manifests.  I've had to switch to helm/helmfile.

Regardless of whether I have the skills to implement such a capability, I don't think the spec for that exists...

**Comment by garry-t** [comment]
should i wait on this nearest time or better look for workarounds? currently deploy oci chart with kustomization and argocd pain in ass.

**Comment by Clockwork-Muse** [comment]
> should i wait on this nearest time or better look for workarounds? currently deploy oci chart with kustomization and argocd pain in ass.

I'm probably not going to have time to deliver this before Jan 1st.

**Comment by porridge** [comment]
Hopefully this would allow signing/verifying the manifests eventually?

**Comment by Clockwork-Muse** [comment]
@porridge - At minimum, the output is going to be a resource that _can_ be signed, as it's going to be a standard OCI image.  Regardless of whether it's something built into kustomize, standard tools can be used to work with the output.

**Comment by Cyper-Madsen** [comment]
@Clockwork-Muse is this still something you will be looking into?

**Comment by Clockwork-Muse** [comment]
@Cyper-Madsen - Yes, I'm still looking into this.
I'm sorry it's taking this long, but I have been working on it.

**Comment by Cyper-Madsen** [comment]
@Clockwork-Muse  sure, just wanted to hear as it's been awhile since last update and you said you switched to helm :)  if you stopped working on it I would give it a go, looks like you have done most of the work already.
